<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GSFA">
<title>Applying GSFA in a simulated dataset • GSFA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applying GSFA in a simulated dataset">
<meta property="og:description" content="GSFA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GSFA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/gsfa_intro.html">Applying GSFA in a simulated dataset</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/xinhe-lab/GSFA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Applying GSFA in a simulated dataset</h1>
                        <h4 data-toc-skip class="author">Yifan Zhou</h4>
            
            <h4 data-toc-skip class="date">2023-10-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/GSFA/blob/HEAD/vignettes/gsfa_intro.Rmd" class="external-link"><code>vignettes/gsfa_intro.Rmd</code></a></small>
      <div class="d-none name"><code>gsfa_intro.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="environment-set-up">Environment set-up<a class="anchor" aria-label="anchor" href="#environment-set-up"></a>
</h2>
<p>Load the <code>GSFA</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/GSFA" class="external-link">GSFA</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">46568</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulate-a-data-set">Simulate a data set<a class="anchor" aria-label="anchor" href="#simulate-a-data-set"></a>
</h2>
<p>We generate a normal gene expression matrix <span class="math inline">\(Y\)</span> with 400 samples and 600 genes and a
binary perturbation matrix <span class="math inline">\(G\)</span> with 2
types of perturbations according to:</p>
<p><span class="math display">\[G_{im} \overset{i.i.d.}{\sim}
\text{Bern}(0.2),
\phi_{ik} \overset{i.i.d.}{\sim} N(0,0.5) \Rightarrow Z = G \beta +
\Phi,\]</span> <span class="math display">\[F_{jk}
\overset{i.i.d.}{\sim} \text{Bern}(0.1),
U_{jk} \overset{i.i.d.}{\sim} N(0, 0.5) \Rightarrow W_{jk}=F_{jk}\cdot
U_{jk},\]</span> <span class="math display">\[E_{ij}
\overset{i.i.d.}{\sim} N(0,1) \Rightarrow Y = ZW^T+E.\]</span> Gene
expression <span class="math inline">\(Y\)</span> was generated from 5
factors, with each factor has ~0.1 of all genes with non-zero loading in
it.</p>
The true association effects between factors and perturbations, <span class="math inline">\(\beta\)</span>, are set to:
<span class="math display">\[\begin{pmatrix}
1 &amp; 0 &amp; 0 &amp; 0 &amp; 0 \\
0 &amp; 0.8 &amp; 0 &amp; 0 &amp; 0
\end{pmatrix}\]</span>
<p>That is, the 1st factor is associated with perturbation 1, and the
2nd factor associated with perturbation 2.</p>
<p>N: Number of samples, P: Number of genes, K: Number of factors, M:
Number of perturbations in the simulated data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_true</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.8</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normal_data_sim.html">normal_data_sim</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">400</span>, P <span class="op">=</span> <span class="fl">600</span>, K <span class="op">=</span> <span class="fl">5</span>, M <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                            beta_true <span class="op">=</span> <span class="va">beta_true</span>,</span>
<span>                            pi_true <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>                            psi_true <span class="op">=</span> <span class="fl">0.5</span>, G_prob <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-gsfa">Fit GSFA<a class="anchor" aria-label="anchor" href="#fit-gsfa"></a>
</h2>
<p>Now we perform GSFA on the given normal expression data and binary
perturbation matrix using <code><a href="../reference/fit_gsfa_multivar.html">fit_gsfa_multivar()</a></code>.</p>
<p>Below, 5 factors are specified in the model, Gibbs sampling is
initialized with truncated SVD for 1000 iterations, with the posterior
means computed using the last 500 iterations.</p>
<p><code>prior_w_s</code>, <code>prior_w_r</code> are prior parameters
of the gene loading on the factors, <code>prior_beta_s</code>, and
<code>prior_beta_r</code> are prior parameters of the effects of
perturbations on the factors. For details about the GSFA model and prior
specification, please see the GSFA paper and Supplementary Notes
(Section 1).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_gsfa_multivar.html">fit_gsfa_multivar</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Y</span>, G <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">G</span>,</span>
<span>                          K <span class="op">=</span> <span class="fl">5</span>, init.method <span class="op">=</span> <span class="st">"svd"</span>,</span>
<span>                          prior_w_s <span class="op">=</span> <span class="fl">10</span>, prior_w_r <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                          prior_beta_s <span class="op">=</span> <span class="fl">5</span>, prior_beta_r <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                          niter <span class="op">=</span> <span class="fl">1000</span>, used_niter <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                          verbose <span class="op">=</span> <span class="cn">T</span>, return_samples <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code>Initializing Z and W with SVD.
Iteration [50] finished.
Iteration [100] finished.
Iteration [150] finished.
Iteration [200] finished.
Iteration [250] finished.
Iteration [300] finished.
Iteration [350] finished.
Iteration [400] finished.
Iteration [450] finished.
Iteration [500] finished.
Iteration [550] finished.
Iteration [600] finished.
Iteration [650] finished.
Iteration [700] finished.
Iteration [750] finished.
Iteration [800] finished.
Iteration [850] finished.
Iteration [900] finished.
Iteration [950] finished.
Iteration [1000] finished.
Computing total effects for each perturbation-gene pair.
Computing posterior means of parameters.</code></pre>
<p>If the option <code>return_samples = T</code>, one can inspect the
traces of samples throughout the iterations stored in
<code>fit0$*_samples</code> slots.</p>
<p>If the sampling chain does not seem to have converged, we can
continue the Gibbs sampling from the previous run by passing the
previous fit object <code>fit0</code> to
<code><a href="../reference/fit_gsfa_multivar.html">fit_gsfa_multivar()</a></code>. Below, the sampling is resumed for
another 1000 iterations, with the posterior means recomputed using the
last 500 iterations.</p>
<p>(Note that we no longer need to specify <code>k</code>,
<code>init.method</code> and prior values this time.)</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_gsfa_multivar.html">fit_gsfa_multivar</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">Y</span>, G <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">G</span>,</span>
<span>                         fit0 <span class="op">=</span> <span class="va">fit0</span>,</span>
<span>                         niter <span class="op">=</span> <span class="fl">1000</span>, used_niter <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="cn">F</span>, return_samples <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<pre><code>Computing total effects for each perturbation-gene pair.
Computing posterior means of parameters.</code></pre>
</div>
<div class="section level2">
<h2 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h2>
<div class="section level3">
<h3 id="perturbation-effects-on-factors">Perturbation effects on factors<a class="anchor" aria-label="anchor" href="#perturbation-effects-on-factors"></a>
</h3>
<p>Note that factors are interchangeable, so their orders won’t
necessarily match the original.</p>
<p>The estimated associations between factors and perturbations are:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">posterior_means</span><span class="op">$</span><span class="va">beta_pm</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">posterior_means</span><span class="op">$</span><span class="va">beta_pm</span><span class="op">)</span>, <span class="op">]</span>,</span>
<span>       digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code>  Factor_1 Factor_2 Factor_3 Factor_4 Factor_5
1 -0.00112 -0.00274  -1.0200 -0.00343 -0.01090
2 -0.00768  0.77100  -0.0148  0.00630 -0.00176</code></pre>
<p>Factor 3 is associated with perturbation 1 with an absolute effect
size of ~1.<br>
Factor 2 is associated with perturbation 2 with an absolute effect size
of ~0.8.</p>
<p>The PIPs (posterior inclusion probability, a measurement of
certainty) of associations between factors and perturbations are:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">posterior_means</span><span class="op">$</span><span class="va">Gamma_pm</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">posterior_means</span><span class="op">$</span><span class="va">Gamma_pm</span><span class="op">)</span>, <span class="op">]</span>,</span>
<span>       digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code>  Factor_1 Factor_2 Factor_3 Factor_4 Factor_5
1     0.04    0.042    1.000    0.032    0.080
2     0.07    1.000    0.096    0.060    0.048</code></pre>
<p>Associations with high certainty are:<br>
Factor 3 ~ Perturbation 1 and Factor 2 ~ Perturbation 2.</p>
<p>Visualization of perturbation effects on factors:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dotplot_beta_PIP.html">dotplot_beta_PIP</a></span><span class="op">(</span><span class="va">fit</span>, target_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturbation 1"</span>, <span class="st">"Perturbation 2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="gsfa_intro_files/figure-html/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="factor-interpretation">Factor interpretation<a class="anchor" aria-label="anchor" href="#factor-interpretation"></a>
</h3>
<p>Genes with non-zero loading on factors can be obtained by
thresholding the gene PIP.</p>
<p>For example, non-zero genes in factor 2 are:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">est_genes_factor2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">posterior_means</span><span class="op">$</span><span class="va">F_pm</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.95</span><span class="op">)</span></span></code></pre></div>
<p>Compare with genes truly in factor 2:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_genes_factor2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">num_olap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">est_genes_factor2</span>, <span class="va">true_genes_factor2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Sensitivity:</span></span>
<span><span class="va">sens</span> <span class="op">&lt;-</span> <span class="va">num_olap</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">true_genes_factor2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sensitivity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">sens</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Specificity:</span></span>
<span><span class="va">fpr</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">est_genes_factor2</span><span class="op">)</span> <span class="op">-</span> <span class="va">num_olap</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Specificity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">fpr</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "Sensitivity: 0.8"
[1] "Specificity: 1"</code></pre>
</div>
<div class="section level3">
<h3 id="perturbation-effects-on-genes">Perturbation effects on genes<a class="anchor" aria-label="anchor" href="#perturbation-effects-on-genes"></a>
</h3>
<p>Differentially expressed genes (DEGs) can be detected by thresholding
LFSR.</p>
<p>DEGs detected under Perturbation 1 and the sensitivity and
specificity of discovery:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_detected1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_detected1</span><span class="op">)</span>, <span class="st">" genes passed LFSR &lt; 0.05."</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">true_genes_factor1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">num_olap1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">genes_detected1</span>, <span class="va">true_genes_factor1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Sensitivity:</span></span>
<span><span class="va">sens1</span> <span class="op">&lt;-</span> <span class="va">num_olap1</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">true_genes_factor1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sensitivity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">sens1</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Specificity:</span></span>
<span><span class="va">fpr1</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_detected1</span><span class="op">)</span> <span class="op">-</span> <span class="va">num_olap1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Specificity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">fpr1</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "46 genes passed LFSR &lt; 0.05."
[1] "Sensitivity: 0.852"
[1] "Specificity: 1"</code></pre>
<p>DEGs detected under Perturbation 2 and the sensitivity and
specificity of discovery:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_detected2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">lfsr</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_detected2</span><span class="op">)</span>, <span class="st">" genes passed LFSR &lt; 0.05."</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">true_genes_factor2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">num_olap2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">genes_detected2</span>, <span class="va">true_genes_factor2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Sensitivity:</span></span>
<span><span class="va">sens2</span> <span class="op">&lt;-</span> <span class="va">num_olap2</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">true_genes_factor2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sensitivity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">sens2</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Specificity:</span></span>
<span><span class="va">fpr2</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_detected2</span><span class="op">)</span> <span class="op">-</span> <span class="va">num_olap2</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="cn">F</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Specificity: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">fpr2</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "44 genes passed LFSR &lt; 0.05."
[1] "Sensitivity: 0.8"
[1] "Specificity: 1"</code></pre>
<p>Visualization of the total effects each perturbation has on selected
genes:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dotplot_total_effect.html">dotplot_total_effect</a></span><span class="op">(</span><span class="va">fit</span>,</span>
<span>                     gene_indices <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">12</span>, <span class="fl">14</span>, <span class="fl">88</span>, <span class="fl">89</span>, <span class="fl">91</span>, <span class="fl">123</span><span class="op">)</span>,</span>
<span>                     target_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Perturbation 1"</span>, <span class="st">"Perturbation 2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="gsfa_intro_files/figure-html/unnamed-chunk-7-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.4.3 dplyr_1.1.3   GSFA_0.2.8   

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.11       plyr_1.8.8        bslib_0.5.1       compiler_4.2.1   
 [5] pillar_1.9.0      jquerylib_0.1.4   tools_4.2.1       digest_0.6.33    
 [9] gtable_0.3.4      jsonlite_1.8.7    evaluate_0.21     memoise_2.0.1    
[13] lifecycle_1.0.3   tibble_3.2.1      pkgconfig_2.0.3   rlang_1.1.1      
[17] cli_3.6.1         rstudioapi_0.15.0 yaml_2.3.7        pkgdown_2.0.7    
[21] xfun_0.40         fastmap_1.1.1     withr_2.5.1       stringr_1.5.0    
[25] knitr_1.44        desc_1.4.2        generics_0.1.3    fs_1.6.3         
[29] vctrs_0.6.3       sass_0.4.7        systemfonts_1.0.4 grid_4.2.1       
[33] rprojroot_2.0.3   tidyselect_1.2.0  glue_1.6.2        R6_2.5.1         
[37] textshaping_0.3.6 fansi_1.0.4       rmarkdown_2.25    farver_2.1.1     
[41] reshape2_1.4.4    purrr_1.0.2       magrittr_2.0.3    scales_1.2.1     
[45] htmltools_0.5.6   colorspace_2.1-0  labeling_0.4.3    ragg_1.2.5       
[49] utf8_1.2.3        stringi_1.7.12    munsell_0.5.0     cachem_1.0.8     </code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yifan Zhou.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
